### PR TITLE
flake: Use builtins.readdir to generate the set of nixosModules

### PR DESCRIPTION
**Pull request by ldesgoui**
I ended up writing something similar in one of my own projects, thought I'd slap it here in case it's wanted.

The diff of `nix flake show` outputs
```diff
1c1
< github:nixos/nixos-hardware/2d44015779cced4eec9df5b8dab238b9f6312cb2
---
> git+file:///home/ldesgoui/nixos-hardware
4a5,6
>     ├───apple: NixOS module
>     ├───apple-macbook-air: NixOS module
70a73,74
>     ├───dell-xps-sleep-resume-bluetooth: NixOS module
>     ├───dell-xps-sleep-resume-i2c-designware: NixOS module
80a85
>     ├───lenovo-ideapad: NixOS module
84a90
>     ├───lenovo-legion-15ich: NixOS module
85a92
>     ├───lenovo-legion-16ach6h-edid: NixOS module
90a98
>     ├───lenovo-thinkpad-e14: NixOS module
96a105
>     ├───lenovo-thinkpad-l14: NixOS module
99a109
>     ├───lenovo-thinkpad-p1-3th-gen: NixOS module
107a118
>     ├───lenovo-thinkpad-t14-amd: NixOS module
111a123
>     ├───lenovo-thinkpad-t14s-amd: NixOS module
132a145
>     ├───lenovo-thinkpad-x1-6th-gen-QHD: NixOS module
151c164
>     ├───lenovo-thinkpad-z-z13: NixOS module
153a167
>     ├───microsoft-surface: NixOS module
154a169,174
>     ├───microsoft-surface-common-ipts: NixOS module
>     ├───microsoft-surface-common-kernel: NixOS module
>     ├───"microsoft-surface-common-kernel-linux-5.19.17": NixOS module
>     ├───"microsoft-surface-common-kernel-linux-6.0.17": NixOS module
>     ├───"microsoft-surface-common-kernel-linux-6.1.6": NixOS module
>     ├───microsoft-surface-common-surface-control: NixOS module
158a179,182
>     ├───microsoft-surface-surface-go: NixOS module
>     ├───microsoft-surface-surface-go-firmware-ath10k: NixOS module
>     ├───microsoft-surface-surface-laptop-amd: NixOS module
>     ├───microsoft-surface-surface-pro-intel: NixOS module
163a188,189
>     ├───onenetbook-4-goodix-stylus-mastykin: NixOS module
>     ├───panasonic-letsnote-cf-lx4: NixOS module
165a192,193
>     ├───pine64-pinebook-pro-firmware-ap6256-firmware: NixOS module
>     ├───pine64-pinebook-pro-keyboard-updater: NixOS module
```

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -1,187 +1,46 @@
...
+        common-pc-laptop-acpi_call = import ./common/pc/laptop/acpi_call.nix;
+        common-pc-laptop-ssd = import ./common/pc/ssd;
+        lenovo-legion-y530-15ich = import ./lenovo/legion/15ich;
+        lenovo-thinkpad-p1-gen3 = import ./lenovo/thinkpad/p1/3th-gen;
+        lenovo-thinkpad-z13 = import ./lenovo/thinkpad/z/z13;
+        letsnote-cf-lx4 = import ./panasonic/letsnote/cf-lx4;
+        microsoft-surface-go = import ./microsoft/surface/surface-go;
+        microsoft-surface-laptop-amd = import ./microsoft/surface/surface-laptop-amd;
+        microsoft-surface-pro-intel = import ./microsoft/surface/surface-pro-intel;
+        purism-librem-15v3 = import ./purism/librem/15v3;
```
Comment (ldesgoui) [inline flake.nix:43]: This was not auto-generated because `15v3` is a link to `13v3`, there's probably a way to fix that

**Comment by ldesgoui** [comment]
Aside from the few mismatches (`asus/battery.nix` could become `asus/battery/default.nix`, `lenovo/thinkpad/p1/3th-gen` could become `lenovo/thinkkpad/p1/gen3`, etc), there *are* some folders we want to ignore (i.e. `pine64/pinebook-pro/firmware/ap6256-firmware/default.nix`), we could have a sentinel file to remove it from the list of modules, i.e. `pine64/pinebook-pro/firmware/ap6256-firmware/.do-not-derive-flake-nixosModule`.

I suppose another way is to have sentinel files for every folder that we do want, I think it would still be better than having to edit the list in `flake.nix`

**Comment by Mic92** [comment]
This also seems to expose some packages as nixos modules.
I guess they could be moved to some directory like `pkgs` so they would be ignored.
I think some modules like `apple` are also not useful to have.

```diff
--- ./old.txt	2023-02-26 20:12:52.665174424 +0100
+++ new.txt	2023-02-26 20:13:03.126040562 +0100
@@ -1,5 +1,7 @@
 acer-aspire-4810t
 airis-n990
+apple
+apple-macbook-air
 apple-macbook-air-3
 apple-macbook-air-4
 apple-macbook-air-6
@@ -66,6 +68,8 @@
 dell-xps-17-9700-intel
 dell-xps-17-9700-nvidia
 dell-xps-17-9710-intel
+dell-xps-sleep-resume-bluetooth
+dell-xps-sleep-resume-i2c-designware
 framework
 framework-12th-gen-intel
 friendlyarm-nanopc-t4
@@ -76,25 +80,31 @@
 hp-elitebook-2560p
 intel-nuc-8i7beh
 kobol-helios4
+lenovo-ideapad
 lenovo-ideapad-15arh05
 lenovo-ideapad-z510
 lenovo-legion-15ach6
 lenovo-legion-15arh05h
+lenovo-legion-15ich
 lenovo-legion-16ach6h
+lenovo-legion-16ach6h-edid
 lenovo-legion-16ach6h-hybrid
 lenovo-legion-16ach6h-nvidia
 lenovo-legion-16ithg6
 lenovo-legion-y530-15ich
 lenovo-thinkpad
+lenovo-thinkpad-e14
 lenovo-thinkpad-e14-amd
 lenovo-thinkpad-e14-intel
 lenovo-thinkpad-e470
 lenovo-thinkpad-e495
 lenovo-thinkpad-l13
 lenovo-thinkpad-l13-yoga
+lenovo-thinkpad-l14
 lenovo-thinkpad-l14-amd
 lenovo-thinkpad-l14-intel
 lenovo-thinkpad-p1
+lenovo-thinkpad-p1-3th-gen
 lenovo-thinkpad-p1-gen3
 lenovo-thinkpad-p14s-amd-gen2
 lenovo-thinkpad-p16s-amd-gen1
@@ -103,10 +113,12 @@
 lenovo-thinkpad-p52
 lenovo-thinkpad-p53
 lenovo-thinkpad-t14
+lenovo-thinkpad-t14-amd
 lenovo-thinkpad-t14-amd-gen1
 lenovo-thinkpad-t14-amd-gen2
 lenovo-thinkpad-t14-amd-gen3
 lenovo-thinkpad-t14s
+lenovo-thinkpad-t14s-amd
 lenovo-thinkpad-t14s-amd-gen1
 lenovo-thinkpad-t410
 lenovo-thinkpad-t420
@@ -128,6 +140,7 @@
 lenovo-thinkpad-x1
 lenovo-thinkpad-x1-10th-gen
 lenovo-thinkpad-x1-6th-gen
+lenovo-thinkpad-x1-6th-gen-QHD
 lenovo-thinkpad-x1-7th-gen
 lenovo-thinkpad-x1-9th-gen
 lenovo-thinkpad-x1-extreme
@@ -146,21 +159,37 @@
 lenovo-thinkpad-x270
 lenovo-thinkpad-x280
 lenovo-thinkpad-z
+lenovo-thinkpad-z-z13
 lenovo-thinkpad-z13
 lenovo-yoga-6-13ALC6
 letsnote-cf-lx4
+microsoft-surface
 microsoft-surface-common
+microsoft-surface-common-ipts
+microsoft-surface-common-kernel
+microsoft-surface-common-kernel-linux-5.19.17
+microsoft-surface-common-kernel-linux-6.0.17
+microsoft-surface-common-kernel-linux-6.1.6
+microsoft-surface-common-surface-control
 microsoft-surface-go
 microsoft-surface-laptop-amd
 microsoft-surface-pro-3
 microsoft-surface-pro-intel
+microsoft-surface-surface-go
+microsoft-surface-surface-go-firmware-ath10k
+microsoft-surface-surface-laptop-amd
+microsoft-surface-surface-pro-intel
 msi-gl62
 msi-gs60
 nxp-imx8qm-mek
 omen-en00015p
 onenetbook-4
+onenetbook-4-goodix-stylus-mastykin
+panasonic-letsnote-cf-lx4
 pcengines-apu
 pine64-pinebook-pro
+pine64-pinebook-pro-firmware-ap6256-firmware
+pine64-pinebook-pro-keyboard-updater
 purism-librem-13v3
 purism-librem-15v3
 raspberry-pi-2
```

**Comment by ldesgoui** [comment]
>  One problem I see with those automatic module list generation is that it's no longer possible to just open flake.nix and do Ctrl-f until one finds the right module.
> Adding a single line for each new hardware profile is usually not that much work.

That's right, if the benefits aren't interesting enough, feel free to close the PR :) Otherwise I can implement the filtering of packages and other unwanted modules

**Comment by Mic92** [comment]
Let's keep the manual way for now. It makes it easier to reason about interfaces.

**Pull request closed without merging**